// @rpgjs/client/lib/index.js
export { Direction, Control, Input, PrebuiltGui, HookServer, HookClient, RpgPlugin, RpgModule, RpgCommonPlayer as RpgSpriteLogic } from '@rpgjs/common';
export { default as entryPoint } from './clientEntryPoint.js';
export { RpgRenderer } from './Renderer.js';
export { Scene as RpgScene } from './Scene/Scene.js';
export { RpgClientEngine } from './RpgClientEngine.js';
export { Spritesheet } from './Sprite/Spritesheet.js';
export { Sound } from './Sound/Sound.js';
export { Howler as RpgGlobalSound } from 'howler';
export { RpgSound } from './Sound/RpgSound.js';
export * as Presets from './Presets/AnimationSpritesheet.js';
export { Animation } from './Effects/AnimationCharacter.js';
export { Animation as AnimationClass } from './Effects/Animation.js';
export { SceneData } from './Scene/SceneData.js';
export { SceneMap as RpgSceneMap } from './Scene/Map.js';
export { RpgGui } from './Gui/Gui.js';
export { Timeline, Ease } from './Effects/Timeline.js';
export { RpgComponent, RpgComponent as RpgSprite } from './Components/Component.js';
export { KeyboardControls } from './KeyboardControls.js';
export { World, room } from 'simple-room-client';
import { spritesheets } from './Sprite/Spritesheets.js';
import { sounds } from './Sound/Sounds.js';
export const RpgResource = {
    spritesheets,
    sounds
};
export { inject } from './inject.js';
