// @rpgjs/server/lib/Gui/ShopGui.js
import { PrebuiltGui } from '@rpgjs/common';
import { Gui } from './Gui.js';
export class ShopGui extends Gui {
    constructor(player) {
        super(PrebuiltGui.Shop, player);
    }
    open(items) {
        items = items.map(item => {
            const it = new item();
            return {
                price: it.price,
                name: it.name,
                description: it.description,
                id: it.id,
                type: item.type
            };
        });
        this.on('buyItem', ({ id, nb }) => {
            try {
                this.player.buyItem(id, nb);
            }
            catch (err) {
                console.log(err);
            }
        });
        this.on('sellItem', ({ id, nb }) => {
            try {
                this.player.sellItem(id, nb);
            }
            catch (err) {
                console.log(err);
            }
        });
        return super.open({ items }, {
            waitingAction: true,
            blockPlayerInput: true
        });
    }
}
