// @rpgjs/types/lib/Input.js
export var Control;
(function (Control) {
    Control["Action"] = "action";
    Control["Attack"] = "attack";
    Control["Defense"] = "defense";
    Control["Skill"] = "skill";
    Control["Back"] = "back";
    Control[Control["Up"] = 1] = "Up";
    Control[Control["Down"] = 3] = "Down";
    Control[Control["Right"] = 2] = "Right";
    Control[Control["Left"] = 4] = "Left";
})(Control || (Control = {}));
export var Input;
(function (Input) {
    Input["Break"] = "break";
    Input["Backspace"] = "backspace";
    Input["Tab"] = "tab";
    Input["Clear"] = "clear";
    Input["Enter"] = "enter";
    Input["Shift"] = "shift";
    Input["Ctrl"] = "ctrl";
    Input["Alt"] = "alt";
    Input["Pause"] = "pause/break";
    Input["CapsLock"] = "caps lock";
    Input["Escape"] = "escape";
    Input["Conversion"] = "conversion";
    Input["NonConversion"] = "non-conversion";
    Input["Space"] = "space";
    Input["PageUp"] = "page up";
    Input["PageDown"] = "page down";
    Input["End"] = "end";
    Input["Home"] = "home";
    Input[Input["Left"] = 4] = "Left";
    Input[Input["Up"] = 1] = "Up";
    Input[Input["Right"] = 2] = "Right";
    Input[Input["Down"] = 3] = "Down";
    Input["Select"] = "select";
    Input["Print"] = "print";
    Input["Execute"] = "execute";
    Input["PrintScreen"] = "Print Screen";
    Input["Insert"] = "insert";
    Input["Delete"] = "delete";
    Input["Zero"] = "0";
    Input["One"] = "1";
    Input["Two"] = "2";
    Input["Three"] = "3";
    Input["Four"] = "4";
    Input["Five"] = "5";
    Input["Six"] = "6";
    Input["Seven"] = "7";
    Input["Height"] = "8";
    Input["Nine"] = "9";
    Input["Equal"] = "=";
    Input["Semicolon"] = "semicolon (firefox), equals";
    Input["LessThan"] = "<";
    Input["Equals"] = "equals (firefox)";
    Input["Beta"] = "\u00DF";
    Input["At"] = "@";
    Input["A"] = "a";
    Input["B"] = "b";
    Input["C"] = "c";
    Input["D"] = "d";
    Input["E"] = "e";
    Input["F"] = "f";
    Input["G"] = "g";
    Input["H"] = "h";
    Input["I"] = "i";
    Input["J"] = "j";
    Input["K"] = "k";
    Input["L"] = "l";
    Input["M"] = "m";
    Input["N"] = "n";
    Input["O"] = "o";
    Input["P"] = "p";
    Input["Q"] = "q";
    Input["R"] = "r";
    Input["S"] = "s";
    Input["T"] = "t";
    Input["U"] = "u";
    Input["V"] = "v";
    Input["W"] = "w";
    Input["X"] = "x";
    Input["Y"] = "y";
    Input["Z"] = "z";
    Input["SearchKey"] = "Windows Key / Left \u2318 / Chromebook Search key";
    Input["NumPad0"] = "numpad 0";
    Input["NumPad1"] = "numpad 1";
    Input["NumPad2"] = "numpad 2";
    Input["NumPad3"] = "numpad 3";
    Input["NumPad4"] = "numpad 4";
    Input["NumPad5"] = "numpad 5";
    Input["NumPad6"] = "numpad 6";
    Input["NumPad7"] = "numpad 7";
    Input["NumPad8"] = "numpad 8";
    Input["NumPad9"] = "numpad 9";
    Input["Multiply"] = "multiply";
    Input["Add"] = "add";
    Input["Subtract"] = "subtract";
    Input["DecimalPoint"] = "decimal point";
    Input["Divide"] = "divide";
    Input["F1"] = "f1";
    Input["F2"] = "f2";
    Input["F3"] = "f3";
    Input["F4"] = "f4";
    Input["F5"] = "f5";
    Input["F6"] = "f6";
    Input["F7"] = "f7";
    Input["F8"] = "f8";
    Input["F9"] = "f9";
    Input["F10"] = "f10";
    Input["F11"] = "f11";
    Input["F12"] = "f12";
    Input["F13"] = "f13";
    Input["F14"] = "f14";
    Input["F15"] = "f15";
    Input["F16"] = "f16";
    Input["F17"] = "f17";
    Input["F18"] = "f18";
    Input["F19"] = "f19";
    Input["F20"] = "f20";
    Input["F21"] = "f21";
    Input["F22"] = "f22";
    Input["F23"] = "f23";
    Input["F24"] = "f24";
    Input["NumLock"] = "num lock";
    Input["ScrollLock"] = "scroll lock";
    Input["CircumflexAccent"] = "^";
    Input["ExclamationMark"] = "!";
    Input["Hash"] = "#";
    Input["Dollar"] = "$";
    Input["AccentU"] = "\u00F9";
    Input["PageBackward"] = "page backward";
    Input["PageForWard"] = "page forward";
    Input["Star"] = "*";
    Input["DecreaseVolume"] = "decrease volume level";
    Input["IncreaseVolume"] = "increase volume level";
    Input["Next"] = "next";
    Input["Previous"] = "previous";
    Input["Stop"] = "stop";
    Input["PlayPause"] = "play/pause";
    Input["Email"] = "e-mail";
    Input["SemiColon"] = "semi-colon / \u00F1";
    Input["EqualSign"] = "equal sign";
    Input["Comma"] = "comma";
    Input["Dash"] = "dash";
    Input["FowardSlach"] = "forward slash / \u00E7";
    Input["GraveAccent"] = "grave accent / \u00F1 / \u00E6";
    Input["OpenBracket"] = "open bracket";
    Input["BackSlach"] = "back slash";
    Input["CloseBracket"] = "close bracket / \u00E5";
    Input["SingleQuote"] = "single quote / \u00F8";
    Input["BackQuote"] = "`";
    Input["Altgr"] = "altgr";
})(Input || (Input = {}));
