// @rpgjs/server/lib/decorators/map.js
import { componentSchema, RpgPlayer } from '../Player/Player.js';
export function MapData(options) {
    return (target) => {
        target.file = options.file;
        target.id = options.id;
        target.type = 'map';
        target.prototype.name = options.name;
        target.prototype.file = options.file;
        target.prototype.id = options.id;
        target.prototype.sounds = options.sounds;
        target.prototype.lowMemory = options.lowMemory;
        target.prototype.$schema = {};
        if (options.syncSchema) {
            target.prototype.$schema = options.syncSchema;
        }
        if (!target.prototype.$schema.shapes) {
            target.prototype.$schema.shapes = [
                {
                    type: String,
                    name: String,
                    x: Number,
                    y: Number,
                    width: Number,
                    height: Number,
                    properties: {
                        collision: Boolean
                    },
                    polygon: [{ x: Number, y: Number }],
                    rotation: Number,
                    components: [componentSchema]
                }
            ];
        }
        if (!target.prototype.$schema.users) {
            target.prototype.$schema.users = [RpgPlayer.schemas];
        }
        if (!target.prototype.$schema.events) {
            target.prototype.$schema.events = [RpgPlayer.schemas];
        }
        target.prototype._events = options.events;
    };
}
